internal/pkg/conditions/poll.go

```diff
@@ -0,0 +1,86 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package conditions
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+)
+
+// AssertionFunc is called every poll interval until it returns nil.
+type AssertionFunc func(ctx context.Context) error
+
+type pollingCondition struct {
+	lastErrMu  sync.Mutex
+	lastErr    error
+	lastErrSet bool
+
+	assertion         AssertionFunc
+	description       string
+	timeout, interval time.Duration
+}
+
+func (p *pollingCondition) String() string {
+	lastErr := "..."
+
+	p.lastErrMu.Lock()
+	if p.lastErrSet {
+		if p.lastErr != nil {
+			lastErr = p.lastErr.Error()
+		} else {
+			lastErr = "OK"
+		}
+	}
+	p.lastErrMu.Unlock()
+
+	return fmt.Sprintf("%s: %s", p.description, lastErr)
+}
+
+func (p *pollingCondition) Wait(ctx context.Context) error {
```

Comment:
Should we look at expanding the retry package for this? Feels similar, but not sure. Or maybe we use the retry package under the hood?

Reply:
This looks very similar, but `retry` has two issues:

* it doesn't support passing `ctx` down to the function
* it doesn't support canceling retry loop on `ctx`

`Condition` seemed to be more natural choice, as it encapsulates both waiting for something and polling under the hood, and it passes `ctx` and it has description for console output.

So if we expand `retry` to handle `ctx`, we can rewrite this to unify it with `retry`, and plus this code limits execution of one attempt to `interval`, so that we can't have 1 attempt taking all the `timeout`.

Reply:
Awesome, sounds like something we can do in the future.